cosmpy/clients/ledger.py

```diff
@@ -0,0 +1,1029 @@
+            ) from last_exception
+
+        # err_code >0 in case of exceptions inside rust contract
+        err_code = res.tx_response.code  # pylint: disable=E1101
+        return MessageToDict(res), err_code
+
+    def get_balance(self, address: str, denom: str) -> int:
+        """
+        Query funds of address and denom
+
+        :param address: Address to be query
+        :param denom: Denom of coins
+
+        :return: Integer representation of amount
+
+        :raises BroadcastException: When communication with node fails.
+        """
+
+        res = None
+        last_exception: Optional[Exception] = None
+        for _ in range(self.n_total_msg_retries):
+            try:
+                res = self.bank_client.Balance(
+                    QueryBalanceRequest(address=str(address), denom=denom)
+                )
+                if res is not None:
+                    break
+            except Exception as e:  # pylint: disable=W0703
+                last_exception = e
+                _logger.warning("Cannot get balance: {%s}", e)
+                self._sleep(self.msg_retry_interval)
+                continue
+
+        if res is None:
+            raise BroadcastException(
+                f"Getting balance failed after multiple attempts: {last_exception}"
+            )
+
+        return int(res.balance.amount)
+
+    def get_balances(self, address: str) -> List[Coin]:
+        """
+        Query all funds of address
+
+        :param address: Address to be query
+
+        :raises BroadcastException: When communication with node fails.
+
+        :return: List of coins
+        """
+
+        res = None
+        last_exception: Optional[Exception] = None
+        for _ in range(self.n_total_msg_retries):
+            try:
+                res = self.bank_client.AllBalances(
+                    QueryBalanceRequest(address=str(address))
+                )
+                if res is not None:
+                    break
+            except Exception as e:  # pylint: disable=W0703
+                last_exception = e
+                _logger.warning("Cannot get balances: {%s}", e)
+                self._sleep(self.msg_retry_interval)
+                continue
+
+        if res is None:
+            raise BroadcastException(
+                f"Getting balances failed after multiple attempts: {last_exception}"
+            )
+
+        return res.balances
+
+    def refill_wealth_from_faucet(
+        self, addresses: List[str], amount: Optional[int] = None
+    ):
+        """
+        Uses faucet api to refill balance of addresses
+
+        :param addresses: List of addresses to be refilled
+        :param amount: Required amount
+        """
+
+        min_amount_required = amount if amount else 500000000
+
+        for address in addresses:
+            attempts_allowed = 10
+
+            # Retry in case of network issues
+            while attempts_allowed > 0:
```

Comment:
it would be nice if we standarize the way we loop for retries. So far we have: `while` with incrementing and decrementing counters and `for` with `range`s